Add support for Transform2D/3D in `lerp()`

aXu-AP opened the pull request:
Implements godotengine/godot-proposals#10579

aXu-AP committed:
Add support for Transform2D/3D in lerp()

Implements godotengine/godot-proposals#10579

Mickeon commented:
The supported types are listed in the description so that should be updated as well.

fire commented:
@lyuma @TokageItLab Transform2D and Transform3D lerps are not really lerps anymore, but from a usability point of view I have nothing blocking it.

aXu-AP commented:
> Transform2D and Transform3D lerps are not really lerps anymore since it's a composite 

Well, lerp of Vector3 is also a composite of lerp of 3 floats and slerp is a different story altogether so I think transform interpolation goes well in to the mix...
If the user calls lerp(transform), there's no mistaking what the user intents to do.

I now changed the description as well. Also fixed a small factual error in the description (instead of clamping the output, `weight` should be clamped, since not all output types can be clamped).

AThousandShips commented on core/variant/variant_utility.cpp:
```diff
@@ -451,13 +451,15 @@ Variant VariantUtilityFunctions::lerp(const Variant &from, const Variant &to, do
 			r_error.argument = 0;
 			r_error.expected = Variant::NIL;
-			return R"(Argument "from" must be "int", "float", "Vector2", "Vector3", "Vector4", "Quaternion", "Basis, or "Color".)";
+			return R"(Argument "from" must be "int", "float", "Vector2", "Vector3", "Vector4", "Quaternion", "Basis", "Transform2D", "Transform3D" or "Color".)";
```
```suggestion
			return R"(Argument "from" must be "int", "float", "Vector2", "Vector3", "Vector4", "Quaternion", "Basis", "Transform2D", "Transform3D", or "Color".)";
```
Keep the Oxford comma

AThousandShips replied:
See [here](https://docs.godotengine.org/en/latest/contributing/documentation/docs_writing_guidelines.html#use-the-oxford-comma-to-enumerate-anything), it's obligatory in the engine and docs

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -621,13 +621,13 @@
 				[/codeblock]
 				See also [method inverse_lerp] which performs the reverse of this operation. To perform eased interpolation with [method lerp], combine it with [method ease] or [method smoothstep]. See also [method remap] to map a continuous series of values to another.
-				[b]Note:[/b] For better type safety, use [method lerpf], [method Vector2.lerp], [method Vector3.lerp], [method Vector4.lerp], [method Color.lerp], [method Quaternion.slerp] or [method Basis.slerp].
+				[b]Note:[/b] For better type safety, use [method lerpf], [method Vector2.lerp], [method Vector3.lerp], [method Vector4.lerp], [method Quaternion.slerp], [method Basis.slerp], [method Transform2D.interpolate_with], [method Transform3D.interpolate_with] or [method Color.lerp].
```
```suggestion
				[b]Note:[/b] For better type safety, use [method lerpf], [method Vector2.lerp], [method Vector3.lerp], [method Vector4.lerp], [method Color.lerp], [method Quaternion.slerp], [method Basis.slerp], [method Transform2D.interpolate_with], or [method Transform3D.interpolate_with].
```
Same here, also let's keep them grouped by method

aXu-AP replied:
Oh I see now. I thought that `Color` was a concept further away from positions, rotations and transforms. But yes, you can sort it this way around too.

AThousandShips replied:
I'd say regardless of choice of sorting `Color` is closer to `Vector4`, it's essentially a fancier `Vector4`, just like `Quaternion`

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -621,13 +621,13 @@
-				Linearly interpolates between two values by the factor defined in [param weight]. To perform interpolation, [param weight] should be between [code]0.0[/code] and [code]1.0[/code] (inclusive). However, values outside this range are allowed and can be used to perform [i]extrapolation[/i]. If this is not desired, use [method clamp] on the result of this function.
-				Both [param from] and [param to] must be the same type. Supported types: [int], [float], [Vector2], [Vector3], [Vector4], [Color], [Quaternion], [Basis].
+				Linearly interpolates between two values by the factor defined in [param weight]. To perform interpolation, [param weight] should be between [code]0.0[/code] and [code]1.0[/code] (inclusive). However, values outside this range are allowed and can be used to perform [i]extrapolation[/i]. If this is not desired, use [method clampf] to limit [param weight].
+				Both [param from] and [param to] must be the same type. Supported types: [int], [float], [Vector2], [Vector3], [Vector4], [Quaternion], [Basis], [Transform2D], [Transform3D], [Color].
```
```suggestion
				Both [param from] and [param to] must be the same type. Supported types: [int], [float], [Vector2], [Vector3], [Vector4], [Color], [Quaternion], [Basis], [Transform2D], [Transform3D].
```
Let's just add at the end instead of inserting in the middle

fire reviewed (approved):
Your arguments to add Transform2d and Transforms3d makes sense for usability.

akien-mga reviewed (approved):
Makes sense to me.

akien-mga commented:
Thanks!
